gc.md

```diff
@@ -189,53 +226,133 @@ ToSpace starts as unallocated memory.
 
 - now we can allocate E
 
-#### Write Barriers
+#### Considerations
 
-[read](http://jayconrod.com/posts/55/a-tour-of-v8-garbage-collection) *Write barriers: the secret ingredient*
+[watch](https://youtu.be/VhpdsjBUS3g?t=15m30s)
 
-[barrier](http://www.memorymanagement.org/glossary/b.html#term-barrier-1) | [write
-barrier](http://www.memorymanagement.org/glossary/w.html#term-write-barrier) | [read
-barrier](http://www.memorymanagement.org/glossary/r.html#term-read-barrier)
+- every allocation brings us closer to GC pause
```

Comment:
By now Mark-Compact GC is mostly concurrent, Scavenger is parallel, so GC pauses should be small: 1ms - 10ms on average.

The pause Mark-Compact GC depends on how many objects are alive on the heap and how fragmented the heap is. If you observe large pauses (>100ms) I would be interested to take a look at the workload.

Reply:
OK I'll add that info in, thanks.
However the fact still remains that you want to avoid allocating/collecting more objects than necessary even though the penalty isn't as bad anymore right?

Reply:
If it possible without sacrificing code quality then yes. Otherwise, it depends :)
For example, temporary short-living objects that are allocated in the young generation and die on the next Scavenge have almost zero cost. This is because GC does not visit dead objects, it visits only live objects.

Reply:
Also I'm bit unclear WRT to allocating + immediately filling a large Array and performance.
It used to be that you'd want to pre-alloc as long as it had `<100K` items, in the below example function _one_ performed better than _two_.

Is this still true?

```js
function one(items) {
  const result = new Array(items.length)
  for (var i = 0; i < items.length; i++) {
    result[i] = items[i] * 10
  }
  return result
}

function two(items) {
  const result = []
  for (const item of items) {
    result.push(item * 10)
  }
  return result
}
```

Reply:
Included the following detail:

> temporary short-living objects that are allocated in the young generation and die on the next Scavenge have almost zero cost. This is because GC does not visit dead objects, it visits only live objects.

here efe68a1.

This was a bit tricky as I'm trying to keep those considerations short.
The details of scavenging and why objects surviving a Scavenge are much more expensive (copying) is outlined in the sections just above, so I think it should be covered.